react-sdk/src/schema/zod.test.ts

```diff
@@ -0,0 +1,474 @@
+                  name: { type: "string" },
+                  value: { type: "number" },
+                },
+              },
+            },
+          },
+        });
+      });
+
+      it("handles deeply nested reused Zod 3 schemas without any $ref", () => {
+        const addressSchema = z3.object({
+          street: z3.string(),
+          city: z3.string(),
+        });
+
+        const personSchema = z3.object({
+          name: z3.string(),
+          homeAddress: addressSchema,
+          workAddress: addressSchema,
+        });
+
+        const result = handleZodSchemaToJson(personSchema);
+
+        expect(hasKeyDeep(result, "$ref")).toBe(false);
+        expect(result).toMatchObject({
+          type: "object",
+          properties: {
+            name: { type: "string" },
+            homeAddress: {
+              type: "object",
+              properties: {
+                street: { type: "string" },
+                city: { type: "string" },
+              },
+            },
+            workAddress: {
+              type: "object",
+              properties: {
+                street: { type: "string" },
+                city: { type: "string" },
+              },
+            },
+          },
+        });
+      });
+
+      it("handles deeply nested reused Zod 4 schemas without any $ref", () => {
+        const addressSchema = z4.object({
+          street: z4.string(),
+          city: z4.string(),
+        });
+
+        const personSchema = z4.object({
+          name: z4.string(),
+          homeAddress: addressSchema,
+          workAddress: addressSchema,
+        });
+
+        const result = handleZodSchemaToJson(personSchema);
+
+        expect(hasKeyDeep(result, "$ref")).toBe(false);
+        expect(result).toMatchObject({
+          type: "object",
+          properties: {
+            name: { type: "string" },
+            homeAddress: {
+              type: "object",
+              properties: {
+                street: { type: "string" },
+                city: { type: "string" },
+              },
+            },
+            workAddress: {
+              type: "object",
+              properties: {
+                street: { type: "string" },
+                city: { type: "string" },
+              },
+            },
+          },
+        });
+      });
+
+      it("handles triple-nested reused Zod 4 schemas without any $ref", () => {
+        const pointSchema = z4.object({
+          x: z4.number(),
+          y: z4.number(),
+        });
+
+        const lineSchema = z4.object({
+          start: pointSchema,
+          end: pointSchema,
+        });
+
+        const shapeSchema = z4.object({
+          outline: z4.array(lineSchema),
+          boundingBox: z4.object({
+            topLeft: pointSchema,
+            bottomRight: pointSchema,
+          }),
+        });
+
+        const result = handleZodSchemaToJson(shapeSchema);
+
+        expect(hasKeyDeep(result, "$ref")).toBe(false);
+      });
```

Comment:
This is the only test that doesn't include `expect(result).toMatchObject({ /*...*/ })` assuming because it is a large object. Could use `expect(result).toMatchSnapshot()` if you think it is worth including.